### PR TITLE
raise error during rebuild if table is missing

### PR DESCRIPTION
**Pull request by snopoke**
http://manage.dimagi.com/default.asp?223821
@czue @esoergel

### CODE REVIEW

File: corehq/apps/userreports/tests/test_save_errors.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import datetime
```
Comment (dimagimon) [inline corehq/apps/userreports/tests/test_save_errors.py]: F401 'datetime' imported but unused

File: corehq/apps/userreports/sql/__init__.py [hunk 2]
```diff
@@ -1,3 +1,3 @@
-from .adapter import get_indicator_table, IndicatorSqlAdapter, metadata
+from .adapter import get_indicator_table, IndicatorSqlAdapter, metadata, ErrorRaisingIndicatorSqlAdapter
```
Comment (dimagimon) [inline corehq/apps/userreports/sql/__init__.py]: F401 'ErrorRaisingIndicatorSqlAdapter' imported but unused
F401 'IndicatorSqlAdapter' imported but unused
F401 'get_indicator_table' imported but unused
F401 'metadata' imported but unused

**Comment by snopoke** [comment]
@czue just read you're last comment on the ticket. Let me know if you think this is a good approach.

**Comment by czue** [comment]
nice this looks good, though I was surprised you overrode `best_effort_save` in the subclass. could you not just override `handle_exception` with the logic and reraising in the subclass and that way the save logic is preserved?

File: corehq/apps/userreports/sql/adapter.py [hunk 3]
```diff
@@ -87,6 +90,24 @@ def delete(self, doc):
...
+    def best_effort_save(self, doc):
+        try:
+            self.save(doc)
+        except IntegrityError:
+            pass  # can be due to users messing up their tables/data so don't bother logging
+        except ProgrammingError as e:
+            orig = getattr(e, 'orig')
+            if orig:
+                error_code = getattr(orig, 'pgcode')
+                if error_code == '42P01':  # http://www.postgresql.org/docs/9.4/static/errcodes-appendix.html
```
Comment (czue) [inline corehq/apps/userreports/sql/adapter.py]: nice.

File: corehq/apps/userreports/sql/__init__.py (hunk 2 above)
Comment (dimagimon) [inline corehq/apps/userreports/sql/__init__.py]: F401 'ErrorRaisingIndicatorSqlAdapter' imported but unused
F401 'IndicatorSqlAdapter' imported but unused
F401 'get_indicator_table' imported but unused
F401 'metadata' imported but unused

**Comment by snopoke** [comment]
hmm, good point. I only added the `handle_exception` piece towards the end
so didn't think too hard about it. Updated

**Comment by czue** [comment]
that was a fast response. :+1: on :white_check_mark:

File: corehq/apps/userreports/sql/adapter.py [hunk 4]
```diff
@@ -87,6 +90,18 @@ def delete(self, doc):
...
 
 
+class ErrorRaisingIndicatorSqlAdapter(IndicatorSqlAdapter):
+    def handle_exception(self, doc, exception):
+        if isinstance(exception, ProgrammingError):
+            orig = getattr(exception, 'orig')
+            if orig:
+                error_code = getattr(orig, 'pgcode')
+                if error_code == '42P01':  # http://www.postgresql.org/docs/9.4/static/errcodes-appendix.html
+                    raise TableNotFoundWarning
```
Comment (esoergel) [inline corehq/apps/userreports/sql/adapter.py]: :star:

File: corehq/apps/userreports/tests/test_save_errors.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        adapter.drop_table()
+
+        doc = {
+            "_id": '123',
+            "domain": "domain",
+            "doc_type": "CommCareCase",
+            "name": 'bob'
+        }
+        with self.assertRaises(TableNotFoundWarning):
+            adapter.best_effort_save(doc)
```
Comment (esoergel) [inline corehq/apps/userreports/tests/test_save_errors.py]: Illustrative test :+1:

**Pull request merged**